src/base.ts

```diff
@@ -0,0 +1,238 @@
+            return new AirtableError(
+                'AUTHENTICATION_REQUIRED',
+                'You should provide valid api key to perform this operation',
+                statusCode
+            );
+        } else if (statusCode === 403) {
+            return new AirtableError(
+                'NOT_AUTHORIZED',
+                'You are not authorized to perform this operation',
+                statusCode
+            );
+        } else if (statusCode === 404) {
+            return (() => {
+                const message =
+                    body && body.error && body.error.message
+                        ? body.error.message
+                        : 'Could not find what you are looking for';
+                return new AirtableError('NOT_FOUND', message, statusCode);
+            })();
+        } else if (statusCode === 413) {
+            return new AirtableError('REQUEST_TOO_LARGE', 'Request body is too large', statusCode);
+        } else if (statusCode === 422) {
+            return (() => {
+                const type =
+                    body && body.error && body.error.type
+                        ? body.error.type
+                        : 'UNPROCESSABLE_ENTITY';
+                const message =
+                    body && body.error && body.error.message
+                        ? body.error.message
+                        : 'The operation cannot be processed';
+                return new AirtableError(type, message, statusCode);
+            })();
+        } else if (statusCode === 429) {
+            return new AirtableError(
+                'TOO_MANY_REQUESTS',
+                'You have made too many requests in a short period of time. Please retry your request later',
+                statusCode
+            );
+        } else if (statusCode === 500) {
+            return new AirtableError(
+                'SERVER_ERROR',
+                'Try again. If the problem persists, contact support.',
+                statusCode
+            );
+        } else if (statusCode === 503) {
+            return new AirtableError(
+                'SERVICE_UNAVAILABLE',
+                'The service is temporarily unavailable. Please retry shortly.',
+                statusCode
+            );
+        } else if (statusCode >= 400) {
+            return (() => {
+                const type =
+                    body && body.error && body.error.type ? body.error.type : 'UNEXPECTED_ERROR';
+                const message =
+                    body && body.error && body.error.message
+                        ? body.error.message
+                        : 'An unexpected error occurred';
+                return new AirtableError(type, message, statusCode);
+            })();
+        } else {
+            return null;
+        }
+    }
+
+    doCall(tableName) {
+        return this.table(tableName);
+    }
+
+    getId() {
+        return this._id;
+    }
+
+    static createFunctor(airtable: Airtable, baseId: string) {
```

Comment:
Can we define a return type for this that includes all the correct methods and properties? things like `new Airtable().base(someId).table(someTable)` are errors otherwise.

Reply:
@SomeHats I'm working on this now, and something I ran into that I can see only one way around. When I define a type that includes the methods that are added below in the `forEach(['table', 'makeRequest', 'runAction', 'getId'], ...` operation, they are not recognized. When I define them as standalone assignments (statically analyzable), then it all works out. The only issue is that it's not very nice to look at.

Reply:
Here's the type I defined: 

```ts
type AirtableBase = {
    (tableName: any): Table;
    _base: Base;
    getId(): string;
    makeRequest(options: BaseRequestOptions): Promise<Record<string, any>>;
    table(tableName: string): Table;
    tables: any;
};
```

Reply:
what if you turn line 218 into `return baseFn as AirtableBase`? that should let us keep the correct types for the return value and not have to worry too much about how it's constructed.

re: that `AirtableBase` type, a few questions:
- can we replace all the `any`s with more specific types?
- can we get rid of `_base` from the public type definition? if no that's fine
- any idea what's up with `tables`? i don't see it defined on `Base` is it just `undefined`? should we remove it?

Reply:
> what if you turn line 218 into return baseFn as AirtableBase? that should let us keep the correct types for the return value and not have to worry too much about how it's constructed.

Smart—that does the trick for the existing `forEach(['table', 'makeRequest', 'runAction', 'getId'], ...` operation. 

> 
> re: that AirtableBase type, a few questions:
> 
> can we replace all the anys with more specific types?

Sure can!

> can we get rid of _base from the public type definition? if no that's fine

Unfortunately, when I remove that property it produces errors.

> any idea what's up with tables? i don't see it defined on Base is it just undefined? should we remove it?

Digging around and I learned that there is no other reference to this property anywhere else in the codebase—I'm going to nuke it.